File: src/core/event.ts
import { ClientEvents } from "eris";
import { Tofu } from "@/core/tofu";

export type TofuEventAction<Event extends keyof ClientEvents> = (
    tofu: Tofu,
    ...args: ClientEvents[Event]
) => Promise<void>;

export interface TofuEvent<Event extends keyof ClientEvents> {
    config: {
        name: Event;
        type: "on" | "once";
    };
    action: TofuEventAction<Event>;
}

File: core/event.go
package core

import (
	"github.com/disgoorg/disgo/bot"
)

type TofuEventAction interface {
	OnEvent(tofu *Tofu, event bot.Event)
}

type TofuEvent struct {
	Action TofuEventAction
}

type tofuEvent[E bot.Event] struct {
	action func(tofu *Tofu, event E)
}

func (l *tofuEvent[E]) OnEvent(tofu *Tofu, e bot.Event) {
	if event, ok := e.(E); ok {
		l.action(tofu, event)
	}
}

func NewTofuEvent[E bot.Event](action func(tofu *Tofu, event E)) TofuEvent {
	return TofuEvent{
		Action: &tofuEvent[E]{action},
	}
}
